### ISSUE TITLE
Support Multi-tenant and Sharded TSO

### ISSUE THREAD

**Issue by binshi-bing** [tikv/pd#5895]
### Describe your feature request related problem

The current TSO solution is known for its poor scalability and single point of failure. As we are moving to 'Cloud' providing DBaaS and Serverless Computing, supporting multi-tenant is one of primary goals which brings more requirements to the TSO service, including:
1. Horizontal scalability across tenants.
2. High resilience and reliability across tenants, i.e., reduce Blast Radius

### Describe the feature you'd like
By sharding TSO service across tenants, we aim to achieve the following goals:
1. Provide unlimited horizontal scalability across all tenants.
2. In all failure scenarios, there is no single point failure for all the tenants.
3. Provide no-degraded service during rolling update deployment and normal operations.  
4. Keep compatibility with the existing features.

#### Big Picture
<img width="2330" alt="image" src="https://user-images.githubusercontent.com/39923490/216216206-79aa7cf8-e555-4591-9777-ad58204339e9.png">

1. A TSO node can manage multiple tenant groups.
2. Each tenant group is an independent TSO service instance.
3. Each tenant group has multiple replicas which are placed in different failure domains to provide HA.
4. The following utility tools will be provided from command console, control plane and even web admin UI.
     - The balancer --  manages the sharding state of the TSO service via etcd.
     - The cluster membership manager -- manages membership of the TSO nodes.
     - The tenant group membership manager -- manages the mapping of tenant -> tenant group.

For more details, please refer to the RFC (TODO: add link to the RFC).

#### Milestone 1 - (goal) deliver single-group tso microservice (**Completed**)
All work are tracked in [#5836](https://github.com/tikv/pd/issues/5836#issuecomment-1412980551) 

#### Milestone 2 - (goal) code complete for multi-group tso microservice (4/21/2023) (**Completed**)
- [x] (02/10) drive design review; finalize design; write RFC @binshi-bing
- [x] **Develop multi-tenant and sharded TSO Service**
    - [x] (04/17) #6231 @rleungx @binshi-bing
    - [x] (04/21) #6232 @binshi-bing @JmPotato @rleungx  @lhy1024 
    - [x] (04/17) #6233 @lhy1024

#### Milestone 3 - (goal) deploy multi-group tso microservice in dev env (5/5/2023) (Note: May 1st - 3rd, holiday)
- [x] tests
    - [x] Add multi-group tests utilities @JmPotato @rleungx 
    - [x] Stabilize the existing test for single group @binshi-bing @rleungx 
    - [x] [**Review**] Improve more multi-group test coverage @binshi-bing @JmPotato @rleungx 
- [x] Integrate with tso downstream services in multi-keyspace env, including create PD client with the new NewClientWithKeyspaceName interface and bootstrapping keyspace group membership/distribution. @rleungx
- [x] Features
    - [x]  Create generic framework for static loading and dynamically watching a rang in etcd, required by tso service discovery and GetMembers Restful APIs which need to add watch loops for the registered tso servers and the primaries of all keyspace groups. #6390  @lhy1024
    - [x]  Enhance split -- add sanity check requiring non-empty member list in the source group; do keyspace assignment when split is triggered for its first time. #6397 @JmPotato 
    - [x]  Add GetMembers Restful APIs (the first version which has no dependency on generic watch framework)  #6373 @JmPotato
- [x] Cherry-pick multi-group tso feature into pd-cse @lhy1024 

#### Milestone 4 - (goal) deploy multi-group tso microservice in staging env (5/29/2023)
- [x] #6142 
    - [x] Define GetMinTS rpc in TSO service https://github.com/pingcap/kvproto/pull/1112 @binshi-bing 
    - [x] #6421 @binshi-bing 
    - [x] Integrate with TiKV
        - [x] #6488  @binshi-bing 
        - [x] https://github.com/tikv/client-go/pull/801 @rleungx 
        - [x] https://github.com/tidbcloud/kvproto/pull/16 @rleungx
        - [x] https://github.com/pingcap/kvproto/pull/1116 @rleungx
- [x] Integrate with tso downstream services in multi-keyspace env @rleungx @binshi-bing
    - [x] #6482 @rleungx
    - [x] #6476 @binshi-bing
- [x] Work with QA team to increase test coverage @rleungx & others
    - [x] #6439 #6445 @lhy1024 
    - [x] #6448 @binshi-bing
    - [x] #6449 @lhy1024
- [x] Amin/Middleware/Utilities
    - [x] Add Keyspace group membership/distribution commands in pd-ctl @JmPotato @lhy1024
        - [x] #6423 @JmPotato
- [x] Cherry-pick more changes, run tests, apply related fixes @lhy1024 
- [x] known bug fixes @all
- [x] deploy Multi-group V1 to staging env
#### （Milestone 4 landing plan) timeline breakdown
- [x] Single-group V1 (tso microservice) reverted in staging and prod env on June 1st due to forward stuck problem. 
- [x] Single-group V2 (Multi-group framework) entered into on June 16 (inclusive).
- [x] Bake Multi-group Version (Multi-group framework) dev env (**May 24 ~ June 16 (exclusive)**). 
       Serverless service team confirmed that there are existing E2E test cases to cover BR & GCSafePoint scenarios.
       Risk: unexpected compatibility issue 
    - [x] Verify Bootstrap process
        - [x] fix #6518 @rleungx @binshi-bing
    - [x] Setup multi-timeline; validate Service Reliability of multi-timeline; restriction validation. @binshi-bing @lhy1024
        - [x] in QA
        - [x] in dev
- [x] QA Multi-group V1 test (**May 24 ~ June 21  (inclusive)**). The following two work will be in parallel (Risk: Test resource capacity):
    - [x] HA/Reliability 
        - [x] HA during Split; Assignee: QA team @rleungx 
            - [x] bug fixes for keyspace group members(distribution) related issue which could cause split to fail. @rleungx  @lhy1024 
        - [x] Setup multi-timeline; validate Service Reliability of multi-timeline with stress/longhual testing. @binshi-bing 
    - [x] Add split benchmark test  @JmPotato
 - [x] Investigation & Bug fixes
        - [x] TSO server stuck issue caused PD out of service during EKS upgrading #6530 #6529 #6534 @binshi-bing 
        - [x] #6526 @binshi-bing
        - [x] #6549 @binshi-bing
        - [x] #6550 @rleungx @lhy1024 
- [x] Multi-group V1 enters into staging (ETA:June 21)
  - [x] Support scale-out and re-balance.
    - [x] Devised solid plan. @Bin Shi @Hanyang Liu 
    - [x] Complete the implementation
      - [x] #6602 @Hanyang Liu 
      - [x] #6617 @Bin Shi 
  - [x] Support no-risk fallback.
    - [x] provide solid fallback design [pd/issues/6589](https://github.com/tikv/pd/issues/6589). @Haizhi Geng @Bin Shi
    - [x] #6625 @Haizhi Geng

### Describe alternatives you've considered

Please see the "The Alternative Architectures Considered" in the RFC (TODO: add link to the RFC).

**Comment by binshi-bing** [comment]
After internal sync within the PD TSO Working Group, here is the roadmap to land on TSO microservice:

1. Mar. 22th
      The single-keyspace-group version (w/o sharding) is ready for baking in the dev environment.
2. Apr. 3rd
    Code complete for the version which supports multiple keyspace groups and sharding.
    The PD client and Keyspace management API are ready for integration with the serverless tier.3.
3. Apr. 17th
      The multi-keyspace-group version (w/ sharding) is ready for baking in the dev environment.

**Comment by binshi-bing** [comment]
#### Future Work
- [ ] #6525  @binshi-bing 
- [ ] Speed up tso service discovery @binshi-bing
- [ ] Implement keyspace movement between keyspace groups(Tenant Group 1 ... N) @JmPotato @binshi-bing 
- [ ] Add GetMembers Restful APIs (the second version, consistent global state) @binshi-bing 
- [ ] TransferPrimary, GetConfiguration RESTful APIs
- [ ] Add cluster membership management
- [ ] Design and implement weight-based leader election mechanism
- [ ] Support Local TSO
- [ ] Support FollowerProxy for tso

**Comment by binshi-bing** [comment]
Move it to the to-do list

    - [ ] #6442  @lhy1024 @rleungx

**Comment by najisawas** [comment]
@binshi-bing i see the RFC link was never updated, do you have it available for share? thank you!
